### PR TITLE
add jupiter perps solana events

### PR DESCRIPTION
**Pull request by andrewhong5297**
Moving an existing query I use for jupiter perps into spellbook for snowflake datashare purposes

### CODE REVIEW

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    , null as pnl_direction
+    , null as pnl_usd --no pnl on increase
+    , to_base58(bytearray_substring(data,1+227,32)) as owner
+    , to_base58(bytearray_substring(data,1+16,32)) as position_key
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+16+32,1))) as position_side --1 is long, 2 is short
+    , to_base58(bytearray_substring(data,1+16+32+1,32)) as custody_position_key --ties to the token mint of long or short token
+    , to_base58(bytearray_substring(data,1+16+32+1+32,32)) as custody_collateral_key --ties to the token mint of the collateral token
+    , block_slot
+    , block_time
+    , date_trunc('month', block_time) as block_month
```
[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/jupiter/solana/jupiter_solana_perp_events.sql:39]: please apply to all three union'd queries

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    , tx_id
+FROM {{ source('solana','instruction_calls') }}
+WHERE executing_account = 'PERPHjGBqRHArX4DySjwM6UJHiR3sWAatqfdBS2qQJu'
+AND bytearray_substring(data,1+8,8) = 0xf5715534d6bb9984 -- IncreasePosition
+AND tx_success = true
+{% if is_incremental() %}
+AND block_time >= date_trunc('day', now() - interval '1' day)
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:46]:
```suggestion
AND {{ incremental_predicate('block_time') }}
```

Comment (jeff-dude) [reply models/jupiter/solana/jupiter_solana_perp_events.sql:46]: please apply to all 3 union'd queries

File: models/jupiter/solana/jupiter_solana_perp_events_schema.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+models:
+  - name: jupiter_solana_perp_events
+    meta:
+      blockchain: solana
+      contributors: [ilemi]
+    config:
+      tags: ['solana','jupiter','perps']
+    description: >
+        solana perp events
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events_schema.yml:11]: please add unique combo of columns test based on keys:
https://github.com/duneanalytics/spellbook/pull/5630/files#diff-639790f5883efe3bae264729381b3a22e1491e2d3a50e36667e8d60cc98c4a46R9

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    , null as pnl_direction
+    , null as pnl_usd --no pnl on increase
+    , to_base58(bytearray_substring(data,1+227,32)) as owner
+    , to_base58(bytearray_substring(data,1+16,32)) as position_key
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+16+32,1))) as position_side --1 is long, 2 is short
+    , to_base58(bytearray_substring(data,1+16+32+1,32)) as custody_position_key --ties to the token mint of long or short token
+    , to_base58(bytearray_substring(data,1+16+32+1+32,32)) as custody_collateral_key --ties to the token mint of the collateral token
+    , block_slot
+    , block_time
+    , cast(date_trunc('month', block_time) as date) as block_month
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:39]: i'm noticing `block_time` can be null in this source table, is that correct?
if yes, then we may need to reconsider as a partition column, as we can't write nulls

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+219,1))) as pnl_direction --0 means negative, 1 means positive
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+220,8))) / 1e6 as pnl_usd --pnl only on decrease position and liquidation
+    , to_base58(bytearray_substring(data,1+228,32)) as owner
+    , to_base58(bytearray_substring(data,1+16,32)) as position_key
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+16+32,1))) as position_side --1 is long, 2 is short
+    , to_base58(bytearray_substring(data,1+16+32+1,32)) as custody_position_key --ties to the token mint of long or short token
+    , to_base58(bytearray_substring(data,1+16+32+1+32,32)) as custody_collateral_key --ties to the token mint of the collateral token
+    , block_slot
+    , block_time
+    , date_trunc('month', block_time) as block_month
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:74]: please update here too

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    , tx_id
+FROM {{ source('solana','instruction_calls') }}
+WHERE executing_account = 'PERPHjGBqRHArX4DySjwM6UJHiR3sWAatqfdBS2qQJu'
+AND bytearray_substring(data,1+8,8) = 0x409c2b4a6d83107f -- DecreasePosition
+AND tx_success = true
+{% if is_incremental() %}
+AND block_time >= date_trunc('day', now() - interval '1' day)
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:81]: please update here too

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+185,1))) as pnl_direction --0 means negative, 1 means positive
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+186,8))) / 1e6 as pnl_usd --pnl only on decrease position and liquidation
+    , to_base58(bytearray_substring(data,1+194,32)) as owner
+    , to_base58(bytearray_substring(data,1+16,32)) as position_key
+    , bytearray_to_bigint(bytearray_reverse(bytearray_substring(data,1+16+32,1))) as position_side --1 is long, 2 is short
+    , to_base58(bytearray_substring(data,1+16+32+1,32)) as custody_position_key --ties to the token mint of long or short token
+    , to_base58(bytearray_substring(data,1+16+32+1+32,32)) as custody_collateral_key --ties to the token mint of the collateral token
+    , block_slot
+    , block_time
+    , date_trunc('month', block_time) as block_month
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:105]: please update here too

File: models/jupiter/solana/jupiter_solana_perp_events.sql [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    , tx_id
+FROM {{ source('solana','instruction_calls') }}
+WHERE executing_account = 'PERPHjGBqRHArX4DySjwM6UJHiR3sWAatqfdBS2qQJu'
+AND bytearray_substring(data,1+8,8) IN (0x68452084d423bf2f, 0x806547a880485654) --LiquidatePosition, LiquidateFullPosition
+AND tx_success = true
+{% if is_incremental() %}
+AND block_time >= date_trunc('day', now() - interval '1' day)
```
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:112]: please update here too

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 4 above)
Comment (andrewhong5297) [reply models/jupiter/solana/jupiter_solana_perp_events.sql:39]: it is null only in 2021. Does not affect perps which were deployed in 2023

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 5 above)
Comment (andrewhong5297) [reply models/jupiter/solana/jupiter_solana_perp_events.sql:74]: it is null only in 2021. Does not affect perps which were deployed in 2023

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 6 above)
Comment (andrewhong5297) [reply models/jupiter/solana/jupiter_solana_perp_events.sql:81]: it is null only in 2021. Does not affect perps which were deployed in 2023

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 5 above)
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:74]: still not updated to match first query

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 6 above)
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:81]: still not updated to match first query

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 7 above)
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:105]: still not updated to match first query

File: models/jupiter/solana/jupiter_solana_perp_events.sql (hunk 8 above)
Comment (jeff-dude) [inline models/jupiter/solana/jupiter_solana_perp_events.sql:112]: still not updated to match first query

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Comment by andrewhong5297** [comment]
haha my bad for not reading your comments carefully. I think I addressed them correctly this time

**Review by jeff-dude — APPROVED** [review]

**Pull request merged**